server/src/main/java/io/unitycatalog/server/service/CoordinatedCommitsService.java

```diff
@@ -0,0 +1,51 @@
+package io.unitycatalog.server.service;
+
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.server.annotation.ExceptionHandler;
+import com.linecorp.armeria.server.annotation.Post;
+import io.unitycatalog.server.exception.GlobalExceptionHandler;
+import io.unitycatalog.server.model.Commit;
+import io.unitycatalog.server.persist.CommitRepository;
+import io.unitycatalog.server.persist.dao.CommitDAO;
+
+import java.util.List;
+import java.util.UUID;
+
+import static io.unitycatalog.server.handler.CoordinatedCommitsHandler.*;
+
+@ExceptionHandler(GlobalExceptionHandler.class)
+public class CoordinatedCommitsService {
+  public static final CommitRepository COMMIT_REPOSITORY = CommitRepository.getInstance();
+
+  @Post("")
+  public HttpResponse commit(Commit commit) {
+    validate(commit);
+    validateTablePath(commit);
+
+    List<CommitDAO> firstAndLastCommits =
+        COMMIT_REPOSITORY.getFirstAndLastCommits(UUID.fromString(commit.getTableId()));
+    if (firstAndLastCommits.isEmpty()) {
+      handleFirstCommit(commit);
+    } else {
+      if (commit.getCommitInfo() == null) {
+        handleBackfillOnlyCommit(
+            commit.getTableId(),
+            commit.getLatestBackfilledVersion(),
+            firstAndLastCommits.get(0),
+            firstAndLastCommits.get(1));
+      } else {
+        if (firstAndLastCommits.get(1).getIsDisownCommit()) {
+          // TODO: Confirm if we need to handle disown and reboard commits
+          handleReboardCommit();
+        } else {
+          handleNormalCommit(commit, firstAndLastCommits.get(0), firstAndLastCommits.get(1));
+        }
+      }
+    }
+    if (commit.getMetadata() != null) {
+      COMMIT_REPOSITORY.updateTableMetadata(commit);
```

Comment:
Should commit and table metadata be updated in a single transaction?

Reply:
fixed